aten/src/ATen/native/cuda/FractionalMaxPool3d.cu

```diff
@@ -350,40 +233,79 @@ void fractional_max_pool3d_backward_out_cuda_template(
 
 }// namespace
 
-std::tuple<Tensor&, Tensor&> fractional_max_pool3d_out_cuda(const at::Tensor& input,
-   IntArrayRef pool_size,
-   IntArrayRef output_size,
-   const at::Tensor& randomSamples,
-   at::Tensor& output,
-   at::Tensor& indices) {
-   fractional_max_pool3d_out_cuda_template(
-     output,
-     indices,
-     input,
-     pool_size,
-     output_size,
-     randomSamples
-   );
-   return std::tuple<Tensor&, Tensor&>(output, indices);
- }
-
-std::tuple<Tensor, Tensor> fractional_max_pool3d_cuda(
-  const at::Tensor& input,
+TORCH_IMPL_FUNC(fractional_max_pool3d_out_cuda) (
+  const Tensor& input,
   IntArrayRef pool_size,
   IntArrayRef output_size,
-  const at::Tensor& randomSamples) {
-    Tensor output = at::empty({0}, input.options());
-    Tensor indices = at::empty({0}, input.options().dtype(kLong));
-    fractional_max_pool3d_out_cuda_template(
-      output,
-      indices,
-      input,
-      pool_size,
-      output_size,
-      randomSamples
-    );
-    return std::tuple<Tensor, Tensor>(output, indices);
+  const Tensor& randomSamples,
+  const Tensor& output,
+  const Tensor& indices
+) {
+
+  int64_t planeDim = 0;
+  int64_t dimt = 1;
+  int64_t dimh = 2;
+  int64_t dimw = 3;
+
+  int64_t outputT = output_size[0];
+  int64_t outputH = output_size[1];
+  int64_t outputW = output_size[2];
+  int64_t poolSizeT = pool_size[0];
+  int64_t poolSizeH = pool_size[1];
+  int64_t poolSizeW = pool_size[2];
+
+  int64_t ndims = input.ndimension();
+  if (ndims == 5) {
+    planeDim++;
+    dimt++;
+    dimh++;
+    dimw++;
+  }
+
+  /* sizes */
+  int64_t numPlanes = input.size(planeDim);
+  int64_t inputT = input.size(dimt);
+  int64_t inputH = input.size(dimh);
+  int64_t inputW = input.size(dimw);
```

Comment:
This op would be a good candidate to write using `___`: you can use it to precompute intermediates in the meta function, which get passed directly into the impl function (so you don't need to recompute them in both the cpu and cuda impl functions). Can you try using that? You can see an example of its usage with `avg_pool2d`: you specify all of the intermediate values in native_functions.yaml ([here](https://github.com/pytorch/pytorch/blob/928ca95ff00adaa640c38c51859d8352a8cb7e07/aten/src/ATen/native/native_functions.yaml#L9146) - you need to specify which value each intermediate is generated from), and you generate them in the meta function ([here](https://github.com/pytorch/pytorch/blob/928ca95ff00adaa640c38c51859d8352a8cb7e07/aten/src/ATen/native/AveragePool2d.cpp#L85)).

I think your intermediates in the yaml would look something like:
```
- output_size -> outputT, outputH, outputW
- pool_size -> poolSizeT, poolSizeH, poolSizeW
- input -> input_, indices_, output_
```
I'm actually not sure about the precomputing the `input_, indices_, output_` - but it's probably worth a try, since it'd let you avoid a bunch of redundant work across the cpu and cuda kernel. LMK what you think!

Reply:
Ok perfect. Tomorrow I will take a look at it and I will let you know. Thanks!